---
title: "Instructions"
author: "Mathieu Flamand"
date: "6/8/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

*wiggleplotr* is a tool to visualise RNA-seq read overage accross annotated exons. A key feature of *wiggleplotr* is that it is able rescale all introns of a gene to fixed length, making it easier to see differences in read coverage between neighbouring exons that can otherwise be too far away.

I have modified *wiggleplotr* to allow other plotting other type of data, such as bed and bedGraph files, which facilitates visualization of m^6^A data

Although it is available on bioconductor, if you want to use these features, you should install it from my github using devtools:

```{r, "Installation"}
library("devtools") # may need to install if not already there
devtools::install_github("mflamand/wiggleplotr")
```

We can then load the required libraries: 

```{r, message=FALSE,warning=FALSE}
# load libraries

library(GenomicFeatures)
library(biomaRt)
library(wiggleplotr)
library(tidyverse)

```

There are at least 3 useful functions in _wiggleplotR_: 

First, the _plotTranscripts_ function allows you to visualize the structure of all transcripts of a gene. It takes the following three inputs, but only the first one is required:

* `exons` - list of GRanges objects containing the start and end coordinates of exons for each transcript.
* `cdss` - list of GRanges objects containing the start and end coordinates of coding sequence (cds) for each transcript (optional).
* `annotations` - a data frame with at least the following three columns: transcript_id, gene_name and strand (optional).

The _plotCoverage_ function can be used to visualize RNA-seq read coverage across the exons of a gene. 

In addition the the `exons`, `cdss` and `transcript_annotations` parameters required by _plotTranscripts_, _plotCoverage_ also requires a `track_data` data frame containing RNA-seq sample metadata as well as path to the read coverage data in BigWig format. 

The _plotCoverageBedGraph_ function is a modified _plotCoverage_ function that can be used to visualize bedgraph coverage across the exons of a gene.it can easily be addapted to plot the C-to-U % in a DART-seq experiment. 

In addition the the `exons`, `cdss` and `transcript_annotations` parameters required by _plotTranscripts_, _plotCoverage_ and _plotCoverageBedGraph_ also requires a `track_data` data frame containing RNA-seq sample metadata as well as path to the read coverage data in BigWig (or Bedgraph) format. 

### import transcript annotations

In order to use those functions, we need to these import transcript annotations for defining the range to be plotted. This can be done from different sources.

1.  Get a list of all genes and transcripts names from ensembl

```{r}

# for a given ensembl release: 

ensembl102 <- useEnsembl(biomart = 'genes', 
                       dataset = "mmusculus_gene_ensembl",
                       version = 102)

selected_attributes = c("ensembl_transcript_id", "ensembl_gene_id",
                        "external_gene_name", "strand",
                        "gene_biotype", "transcript_biotype")

transcript_metadata = getBM(attributes = selected_attributes, mart = ensembl102)


transcript_metadata = dplyr::rename(transcript_metadata, 
                     transcript_id = ensembl_transcript_id,
                     gene_id = ensembl_gene_id,
                     gene_name = external_gene_name)

# we can view how this table looks like
transcript_metadata

# the object can be saved to avoid fetching the info every time:
 
saveRDS(transcript_metadata, "metadata.rds")

# you can then retrieve it on a new session with : 

transcript_metadata = readRDS("metadata.rds")

```

2. Alternatively, the same information can be retrieve from a gtf file obtained from ensembl or UCSC etc. 

```{r}

#alternatively you can build a similar table from a gtf file:

#this may be easier if you are after NCBI RefSeq or UCSC annotations 

temp<-readr::read_tsv("../data/mm10.ncbiRefSeq2.gtf.gz",col_names =c("chr","source","type","start","end","score","strand","frame","attributes"))

transcript_metadata_ncbi<- temp %>% filter(type=="transcript"|type=="CDS")%>%
  tidyr::separate(.,col="attributes",into=c("gene_id","transcript_id","gene_name"),sep=";") %>% 
  mutate(gene_id=(stringr::str_extract(gene_id, stringr::regex("(?<=\").+(?=\")"))),
         transcript_id=(stringr::str_extract(transcript_id, stringr::regex("(?<=\").+(?=\")"))),
         gene_name=stringr::str_extract(gene_name, stringr::regex("(?<=\").+(?=\")")))

cds_list<-unique((transcript_metadata_ncbi %>% filter(type=="CDS"))$transcript_id)

 transcript_metadata_ncbi<-transcript_metadata_ncbi %>% filter(type=="transcript") %>% mutate(gene_biotype = if_else(transcript_id %in% cds_list, "protein_coding","non_coding")) %>% dplyr::select("transcript_id","gene_id","gene_name","strand","gene_biotype")


 rm(temp,cds_list)

saveRDS(transcript_metadata_ncbi, "metadata_ncbi.rds")
#transcript_metadata_ncbi = readRDS("metadata_ncbi.rds")


```

### import transcript positions

Now that we know what are the transcript_ids for each genes, we need to know what are the exons, introns, and cds of each of these transcripts. We can again do this using 2 different ways:

1. Get a transcript database with the position of all exons for each transcripts through biomaRt:


```{r}

#from biomaRt

mart<-useMart("ENSEMBL_MART_ENSEMBL")

#be sure to find the right server for your desired release using listEnsemblArchives(). you may need to change https to http in the server used

txdb = makeTxDbFromBiomart(biomart = "ENSEMBL_MART_ENSEMBL",
                            dataset = "mmusculus_gene_ensembl",
                             host="http://nov2020.archive.ensembl.org")

# you can save this database for fast loading 
saveDb(txdb, "txdb_file.rds")
txdb = loadDb("txdb_file.rds")
```

2. or get transcript database from the GTF file 

```{r}

# or you can obtain this from the same GTF used in the first step with :

txdb_ncbi <- makeTxDbFromGFF("../data/mm10.ncbiRefSeq2.gtf.gz",organism="Mus musculus")
 
saveDb(txdb_ncbi, "txdb_ncbi_file.rds")
txdb_ncbi = loadDb("txdb_ncbi_file.rds")
```

We can then extract the list of all exons and ORF from those database with the following functions:

```{r}
exons = exonsBy(txdb, by = "tx", use.names = TRUE)
cdss = cdsBy(txdb, by = "tx", use.names = TRUE)

exons_ncbi = exonsBy(txdb_ncbi, by = "tx", use.names = TRUE)
cdss_ncbi = cdsBy(txdb_ncbi, by = "tx", use.names = TRUE)


```


With this information we can now plot the transcript structure for any given gene:

We can plot only the gene structure with plotTranscripts:

```{r}

# we first select which transcript we want to plot from the metadata table: 

# for all transcript of a given gene:
selected_transcripts = transcript_metadata %>%
  dplyr::filter(gene_name == "Actb", transcript_biotype == "protein_coding")

selected_transcripts

# Wt then extracty the Ids of each transcript to be plotted
tx_ids = selected_transcripts$transcript_id

# and we can plot the transcript:

plotTranscripts(exons,
                cdss,
                tx_ids=tx_ids,
                transcript_metadata,
                rescale_introns = FALSE)

# We can also rescale the introns if we want to focus on exons only. For this we turn the option rescale_introns to TRUE and we can also indicate the length with new_intro_length:

plotTranscripts(exons,
                cdss,
                tx_ids=tx_ids,
                transcript_metadata,
                rescale_introns = TRUE,
                ) 
## or with minimal intron size:
plotTranscripts(exons,
                cdss,
                tx_ids=tx_ids,
                transcript_metadata,
                rescale_introns = TRUE,
                new_intron_length = 1,
                
                ) 

# other useful options: 
# new_intron_length = 50      ## change relative size of introns 
# flanking_length = c(50,50)   ## add space on either size of plot
# transcript_label = TRUE    ## change to FALSE to remove labels for each transcripts
```

If we want to plot coverage, we need to define a list of files 

to plot coverage we need to define a list of bigwig files that will be read:

```{r}

# if bigwigs are stranded, you can make a data frame for each set

# we define a dataframe with the sample names, and factors associated with each samples:

sample_data = data_frame(

    sample_id = c("mNH_333", "mNH_334", "mNH_371","mNH_372","mNH_389",
                "mNH_390","mNH_391","mNH_392","mNH_393","mNH_394",
                "mNH_395","mNH_396","mNH_397","mNH_398","mNH_399","mNH_400"),
  
    Genotype = factor(c(rep("KO",4),rep("WT",8),rep("KO",4)), levels = c("WT", "KO")),
  
    Compartment=factor(c(rep(c("Soma","Neurites"),8)), levels = c("Soma","Neurites")),
  
    Condition=factor(c(rep("Ctrl",2),rep("KCl",2), rep("Ctrl",4),rep("KCl",4),rep("Ctrl",2),rep("KCl",2)), levels=c("Ctrl","KCl")),
  
    scaling_factor = 1)

# we then add a column with the path to the files
sample_data = sample_data %>%
  dplyr::mutate(bigWig = paste0("../../../bigiwg_bpm/",sample_id,".bw"))

# We can then see the table with metadata
sample_data

# Finally we add a color group and track id to each sample so we can color them in the coverage track
track_data = dplyr::mutate(sample_data, track_id = paste(Genotype,Compartment,Condition,sep = ":"), colour_group = Compartment)

```

We can also import bed files for visualization (if wanted):

```{r}
MeRIP<-BiocIO::import("D:/OneDrive/OneDrive - Duke University/RNAseq/Flamand_6511/m6a/bed/MeRIP.sorted.bed", format	="bed")

DART<-BiocIO::import(con="D:/OneDrive/OneDrive - Duke University/RNAseq/Flamand_6511/m6a/bed/Neurons_common.bed", format="bed")

# if single nucleotide resolution, you can widen the peaks for better visualization
start(DART) <- start(DART) - 5
end(DART) <- end(DART) + 5

# make a list of all datasets (if you have a single datasets, put it as a list as well)
bed<-list(MeRIP,DART)
#name each element of the list
names(bed)<-c("MeRIP","DART")
```

We can plot only the gene structure and bed files with plotTranscripts by adding the bed object:

```{r}

# for all transcript of a given gene:
selected_transcripts = transcript_metadata %>%
  dplyr::filter(gene_name == "Camk2a" )

selected_transcripts

# Wt then extracty the Ids of each transcript to be plotted
tx_ids = selected_transcripts$transcript_id


plotTranscripts(exons,
                cdss,
                tx_ids=tx_ids,
                transcript_metadata,
                rescale_introns = TRUE,
                new_intron_length = 1,
                ) 

```
with many transcript, it can be hard to make sense of gene tracks: we can select a single transcript to plot:
```{r}

selected_transcripts = transcript_metadata %>%
  dplyr::filter(gene_name == "Camk2a", transcript_id == "ENSMUST00000102888" )

selected_transcripts

# Wt then extracty the Ids of each transcript to be plotted
tx_ids = selected_transcripts$transcript_id


plotTranscripts(exons,
                cdss,
                tx_ids=tx_ids,
                transcript_metadata,
                rescale_introns = TRUE,
                new_intron_length = 1,
                bed_sites = bed
                ) 


```

We can plot the coverage with the plotCoverage() function (from bigwig files)

```{r}

# in addition to the information provided above, we need to provide the metadata table create earlier with the path to bigwigs

plotCoverage(exons=exons, # list of exons
             cdss=cdss,  #list of cdss
             tx_ids=tx_ids,
             transcript_annotations =transcript_metadata, # metadata to label gene name
             track_data = track_data,   
             fill_palette =  c("red","blue") , # colors for each dataset (at least as much as number of conditions)
             alpha = 1,  # transparency
             new_intron_length = 1,  # size of rescaled introns (defaults to 50)
             transcript_label = TRUE, 
             rescale_introns = TRUE,
             coverage_type='line_sd',   # one of: line, area, both, line_sd
             heights=c(0.60,0.40),    # relative size of both plot in final plot (coverage plot and gene structure plot)
             mean_only=TRUE,          # if set to false, each replicate is displayed instead of the average
             plot_fraction = 0.5,     # subsample bigwig for faster plotting
             flanking_length = c(50, 50), 
             return_subplots_list =FALSE,  # return coverage and gene structure plots without combining them together. 
             bed_sites = bed) # list of bed sites

# grouping can be changed on the fly by changing the track_data within the plot, for example:


#would result in the coverage plot in 2 panels (input and RIP separated.)

```

```{r}

# When you have to many conditions, it may be easier to group them to efficiently vizualize them:

# for example, if we want to group them by 2 factors instead of 3, for example (Genotype and compartments here): you can modify the track data to change the track_id field (which determines how the data is faceted)

track_data %>% mutate(track_id = factor(paste(Genotype,Compartment,sep = ":"),levels=(c("WT:Soma","WT:Neurites","KO:Soma","KO:Neurites")))) %>% 
  plotCoverage(exons=exons, # list of exons
             cdss=cdss,  #list of cdss
             tx_ids=tx_ids,
             transcript_annotations =transcript_metadata, # metadata to label gene name
             track_data = .,   
             fill_palette =  c("blue","red") , # colors for each dataset (at least as much as number of conditions)
             alpha = 1,  # transparency
             new_intron_length = 25,  # size of rescaled introns (defaults to 50)
             transcript_label = TRUE, 
             rescale_introns = TRUE,
             coverage_type='line_sd',   # one of: line, area, both, line_sd
             heights=c(0.60,0.40),    # relative size of both plot in final plot (coverage plot and gene structure plot)
             mean_only=TRUE,          # if set to false, each replicate is displayed instead of the average
             plot_fraction = 0.5,     # subsample bigwig for faster plotting
             flanking_length = c(50, 50), 
             return_subplots_list =FALSE,  # return coverage and gene structure plots without combining them together. 
             bed_sites = bed) # list of bed sites

# We can also change how the track is displayed with the coverage_type: choose one of: "line", "area", "both", "line_sd"

track_data %>% mutate(track_id = factor(paste(Genotype,Compartment,sep = ":"),levels=(c("WT:Soma","WT:Neurites","KO:Soma","KO:Neurites")))) %>% 
  plotCoverage(exons=exons, # list of exons
             cdss=cdss,  #list of cdss
             tx_ids=tx_ids,
             transcript_annotations =transcript_metadata, # metadata to label gene name
             track_data = .,   
             fill_palette =  c("blue","red") , # colors for each dataset (at least as much as number of conditions)
             alpha = 1,  # transparency
             new_intron_length = 25,  # size of rescaled introns (defaults to 50)
             transcript_label = TRUE, 
             rescale_introns = TRUE,
             coverage_type='both',   # one of: line, area, both, line_sd
             heights=c(0.60,0.40),    # relative size of both plot in final plot (coverage plot and gene structure plot)
             mean_only=TRUE,          # if set to false, each replicate is displayed instead of the average
             plot_fraction = 0.5,     # subsample bigwig for faster plotting
             flanking_length = c(50, 50), 
             return_subplots_list =FALSE,  # return coverage and gene structure plots without combining them together. 
             bed_sites = bed) # list of bed sites

# individual replicates can also be displayed by turning the mean_only option to false:

track_data %>% mutate(track_id = factor(paste(Genotype,Compartment,sep = ":"),levels=(c("WT:Soma","WT:Neurites","KO:Soma","KO:Neurites")))) %>% 
  plotCoverage(exons=exons, # list of exons
             cdss=cdss,  #list of cdss
             tx_ids=tx_ids,
             transcript_annotations =transcript_metadata, # metadata to label gene name
             track_data = .,   
             fill_palette =  c("blue","red") , # colors for each dataset (at least as much as number of conditions)
             alpha = 1,  # transparency
             new_intron_length = 1,  # size of rescaled introns (defaults to 50)
             transcript_label = TRUE, 
             rescale_introns = TRUE,
             coverage_type='line',   # one of: line, area, both, line_sd
             heights=c(0.60,0.40),    # relative size of both plot in final plot (coverage plot and gene structure plot)
             mean_only=FALSE,          # if set to false, each replicate is displayed instead of the average
             plot_fraction = 0.05,     # subsample bigwig for faster plotting
             flanking_length = c(50, 50), 
             return_subplots_list =FALSE,  # return coverage and gene structure plots without combining them together. 
             bed_sites = bed) # list of bed sites

```

If a transcript is on the negative strand, it will be reverse by default

```{r}

selected_transcripts = transcript_metadata_ncbi %>%
  dplyr::filter(gene_name == "Actb")

selected_transcripts

# Wt then extracty the Ids of each transcript to be plotted
tx_ids = selected_transcripts$transcript_id


track_data %>% mutate(track_id = factor(paste(Genotype,Compartment,sep = ":"),levels=(c("WT:Soma","WT:Neurites","KO:Soma","KO:Neurites")))) %>% 
  plotCoverage(exons=exons_ncbi, # list of exons
             cdss=cdss_ncbi,  #list of cdss
             tx_ids=tx_ids,
             transcript_annotations =transcript_metadata_ncbi, # metadata to label gene name
             track_data = .,   
             fill_palette =  c("blue","red") , # colors for each dataset (at least as much as number of conditions)
             alpha = 1,  # transparency
             new_intron_length = 1,  # size of rescaled introns (defaults to 50)
             transcript_label = TRUE, 
             rescale_introns = TRUE,
             coverage_type='both',   # one of: line, area, both, line_sd
             heights=c(0.60,0.40),    # relative size of both plot in final plot (coverage plot and gene structure plot)
             mean_only=TRUE,          # if set to false, each replicate is displayed instead of the average
             plot_fraction = 0.5,     # subsample bigwig for faster plotting
             return_subplots_list =FALSE,  # return coverage and gene structure plots without combining them together.
             bed_sites=bed) # list of bed sites


```
You can turn off this off with : no_reverse=TRUE


```{r}

track_data %>% mutate(track_id = factor(paste(Genotype,Compartment,sep = ":"),levels=(c("WT:Soma","WT:Neurites","KO:Soma","KO:Neurites")))) %>% 
  plotCoverage(exons=exons_ncbi, # list of exons
             cdss=cdss_ncbi,  #list of cdss
             tx_ids=tx_ids,
             transcript_annotations =transcript_metadata_ncbi, # metadata to label gene name
             track_data = .,   
             fill_palette =  c("blue","red") , # colors for each dataset (at least as much as number of conditions)
             alpha = 1,  # transparency
             new_intron_length = 1,  # size of rescaled introns (defaults to 50)
             transcript_label = TRUE, 
             rescale_introns = TRUE,
             coverage_type='both',   # one of: line, area, both, line_sd
             heights=c(0.60,0.40),    # relative size of both plot in final plot (coverage plot and gene structure plot)
             mean_only=TRUE,          # if set to false, each replicate is displayed instead of the average
             plot_fraction = 0.5,     # subsample bigwig for faster plotting
             return_subplots_list =FALSE,  # return coverage and gene structure plots without combining them together. 
             no_reverse = TRUE,
             bed_sites = bed
             ) # list of bed sites


```


### Plotting bedgraph files

The principle is the same for bedgraph files: we need to provide a table with the path to the bedgraph files

For example: 

```{r}

sample_data <- data_frame(
  sample_id = c("WT", "KO"),
  sample = factor(c("WT", "KO"), levels = c("WT", "KO")), 
  scaling_factor = 1) 

sample_data <- sample_data %>%
  dplyr::mutate(bedgraph = paste0("./",sample_id,".bedgraph"))  # change for position of files

track_data <- dplyr::mutate(sample_data, track_id = sample, colour_group = sample)


```

Now we can plot using the plotCoverageBedGraph function with very similar options. However, this function does not average replicates. you will need to compute this prior to plotting the data.

```{r}

selected_transcripts = transcript_metadata %>%
  dplyr::filter(gene_name == "Ntrk2",transcript_biotype=="protein_coding",(transcript_id=="ENSMUST00000079828" | transcript_id =="ENSMUST00000109838"))

selected_transcripts

# Wt then extracty the Ids of each transcript to be plotted
tx_ids = selected_transcripts$transcript_id


plotCoverageBedGraph(exons,
                     cdss,
                     fill_palette = c("black","red"),
                     tx_ids=tx_ids,
                     transcript_metadata,
                     track_data = track_data,
                     new_intron_length = 1, 
                     transcript_label = TRUE,
                     rescale_introns = TRUE,
                     return_subplots_list = FALSE,
                    )
```


### advanced options

gene structure and coverage plots can also be returned as individual panels to change aesthetics for final plotting:

We can define theme functions for each subplots (gene track and gene structure) 
```{r}

theme_a <- function( ){
  theme(#panel.grid.major.y = element_line(color="gray95", size=0.125),
        axis.text.y = element_text(size=4, color="black", hjust=1,margin = margin(r = 0.1)),
        axis.ticks.length.y = unit(.1, "line"),
        panel.border = element_rect(color="black", size=0.25),
        strip.background = element_blank(),
        strip.text.y = element_blank(),
        axis.title = element_text(size=6, color="black",margin = margin(r = 0.5)),
        axis.ticks.y = element_line(color="black",size=0.25),
        axis.line.y = element_line(color="black",size=0.25),
        plot.margin=unit(c(0,0,0,0),"line"), panel.spacing = unit(0.2,"line")
        )
}

theme_b <- function(){
   theme(panel.border = element_rect(color="black",size=0.25),
        axis.text.x = element_text(size=5, color="black",margin = margin(t = 0)),
        axis.title.x = element_text(size=6, color="black",margin = margin(t = 0.5)), 
        axis.ticks.length.x = unit(.1, "line"),
        strip.background = element_blank(),
        strip.text = element_blank(),
        axis.ticks.x = element_line(color="black"),
        plot.margin=unit(c(-0.1,0,0,0),"line"))
}
```

when plotting bigwigs or bedgraph, you can use the return_subplots_list option to return individual plots that can be edited with ggplot

```{r}
plots<-plotCoverageBedGraph(exons,
                     cdss,
                     fill_palette = c("black","red"),
                     tx_ids=tx_ids,
                     transcript_metadata,
                     track_data = track_data,
                     new_intron_length = 1, 
                     transcript_label = FALSE,
                     rescale_introns = TRUE,
                     return_subplots_list = TRUE,
                    )
plots[[1]]

plots[[2]]

```

```{r, fig.height=1.4, fig.width=2.4}
p1<-plots[[1]]+
  theme_a()+
  scale_y_continuous(labels = scales::percent)
 
p1 


p2<-plots[[2]]+labs(x="Distance from mRNA start (nt)")+
  theme_b()
p2

# and we can rebuild the final plot using cowplots

cowplot::plot_grid(p1,p2, rel_heights =  c(2,1), nrow=2, align = "v",labels = "Ntrk2",label_size=6,label_fontface="italic",label_x=0.8)


# these plots can then be saved with the ggsave() function
```


